### PR TITLE
ES|QL: Fix RLIKE folding with (unsupported) case insensitive pattern

### LINKED ISSUE #118371: ES|QL: inconsistent behavior with RLIKE and case insensitive regex syntax
RLIKE operator, with a case insensitive regex (eg. `(?i)foo`) behaves differently when being folded compared to normal execution.

In particular, the case insensitive prefix `(?i)` is considered and used only when the expression can be folded

Schema:
```
POST /test
{
    "mappings": {
        "properties":{
            "action": {
                "type":"keyword"
            }
        }
    }
}


```

```
PUT test/_doc/1
{
  "action": "StopInstance"
}
```

Normal regex query: returns the record as expected

```
POST /_query
{
  "query": """
FROM rlike_test 
| WHERE action RLIKE "Stop.*"
  """
}
```

Query with case insensitive syntax `(?i)`: returns nothing

```
POST /_query
{
  "query": """
FROM rlike_test 
| WHERE action RLIKE "(?i)Stop.*"
  """
}
```

Query with ROW (folded expression) and case insensitive syntax `(?i)`: returns one record

```
POST /_query
{
  "query": """
ROW action = "StopInstance"
| WHERE action RLIKE "(?i)Stop.*"
  """
}
```

### PR DESCRIPTION
**Pull request by luigidellaquila**
RLIKE with `(?i)` prefix (case-insensitive) is not supported in ES|QL, but in a specific case (constant folding) the calculation relied on Java regex, that actually supports it, making the overall behavior inconsistent.

With this fix, `(?i)` prefix is not taken into consideration as a case-insensitive option.

Taking a chance to cleanup a bit the `RegexMatch` operator hierarchy.

Fixes: https://github.com/elastic/elasticsearch/issues/118371

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/RLike.java [hunk 1]
```diff
@@ -93,6 +93,11 @@ protected TypeResolution resolveType() {
         return isString(field(), sourceText(), DEFAULT);
     }
 
+    @Override
+    public Boolean fold() {
+        return (Boolean) EvaluatorMapper.super.fold();
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/RLike.java:98]: This is the actual fix, all the rest is refactoring and code cleanup

**Review by ivancea — APPROVED** [review]
LGTM!

**Comment by nik9000** [comment]
I'm also backporting to 8.17.1: https://github.com/elastic/elasticsearch/pull/119292

**Comment by luigidellaquila** [comment]
Thanks @nik9000!

**Comment by nik9000** [comment]
> Thanks @nik9000!

Happy to do it.

**Pull request merged**